src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticQueue.cs

```diff
@@ -0,0 +1,211 @@
+                        Interlocked.CompareExchange(ref _localSemanticDiagnostics, new ConcurrentDictionary<DiagnosticAnalyzer, SimpleDiagnosticQueue>(), null);
+                    }
+
+                    EnqueueCore(_localSemanticDiagnostics, diagnostic, analyzer);
+                }
+            }
+
+            public override void EnqueueNonLocal(Diagnostic diagnostic, DiagnosticAnalyzer analyzer)
+            {
+                if (_nonLocalDiagnostics == null)
+                {
+                    Interlocked.CompareExchange(ref _nonLocalDiagnostics, new ConcurrentDictionary<DiagnosticAnalyzer, SimpleDiagnosticQueue>(), null);
+                }
+
+                EnqueueCore(_nonLocalDiagnostics, diagnostic, analyzer);
+            }
+
+            private static void EnqueueCore(ConcurrentDictionary<DiagnosticAnalyzer, SimpleDiagnosticQueue> diagnosticsMap, Diagnostic diagnostic, DiagnosticAnalyzer analyzer)
+            {
+                diagnosticsMap.AddOrUpdate(analyzer,
+                    new SimpleDiagnosticQueue(diagnostic),
+                    (a, queue) =>
+                    {
+                        queue.Enqueue(diagnostic);
+                        return queue;
+                    });
+            }
+
+            public override bool TryComplete()
+            {
+                return true;
+            }
+
+            private static bool TryDequeue(ConcurrentDictionary<DiagnosticAnalyzer, SimpleDiagnosticQueue> diagnosticsMap, out Diagnostic d)
+            {
+                Diagnostic diag = null;
+                if (diagnosticsMap != null && diagnosticsMap.Any(kvp => kvp.Value.TryDequeue(out diag)))
+                {
+                    d = diag;
+                    return true;
+                }
+
+                d = null;
+                return false;
+            }
+
+            public override bool TryDequeue(out Diagnostic d)
+            {
+                return TryDequeue(_localSemanticDiagnostics, out d) ||
+                    TryDequeue(_localSyntaxDiagnostics, out d) ||
+                    TryDequeue(_nonLocalDiagnostics, out d);
+            }
+
+            public override ImmutableArray<Diagnostic> GetLocalSyntaxDiagnostics(DiagnosticAnalyzer analyzer)
+            {
+                return GetDiagnosticsCore(analyzer, _localSyntaxDiagnostics);
+            }
+
+            public override ImmutableArray<Diagnostic> GetLocalSemanticDiagnostics(DiagnosticAnalyzer analyzer)
+            {
+                return GetDiagnosticsCore(analyzer, _localSemanticDiagnostics);
+            }
+
+            public override ImmutableArray<Diagnostic> GetNonLocalDiagnostics(DiagnosticAnalyzer analyzer)
+            {
+                return GetDiagnosticsCore(analyzer, _nonLocalDiagnostics);
+            }
+
+            private ImmutableArray<Diagnostic> GetDiagnosticsCore(DiagnosticAnalyzer analyzer, ConcurrentDictionary<DiagnosticAnalyzer, SimpleDiagnosticQueue> diagnosticsMap)
```

Comment:
I don't think this is GetXXX since it has side effect. 2 calls return 2 different result. how about DequeueXXX or TakeXXX